frame/fast-unstake/src/lib.rs

```diff
@@ -0,0 +1,500 @@
+//! to prevent them from accidentally exposing themselves behind a validator etc.
+//!
+//! Once processed, if successful, no additional fees for the checking process is taken, and the
+//! staker is instantly unbonded. Optionally, if the have asked to join a pool, their *entire* stake
+//! is joined into their pool of choice.
+//!
+//! If unsuccessful, meaning that the staker was exposed sometime in the last `BondingDuration` eras
+//! they will end up being slashed for the amount of wasted work they have inflicted on the chian.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use pallet::*;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+// NOTE: enable benchmarking in tests as well.
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+mod types;
+pub mod weights;
+
+pub const LOG_TARGET: &'static str = "runtime::fast-unstake";
+
+// syntactic sugar for logging.
+#[macro_export]
+macro_rules! log {
+	($level:tt, $patter:expr $(, $values:expr)* $(,)?) => {
+		log::$level!(
+			target: crate::LOG_TARGET,
+			concat!("[{:?}] 💨 ", $patter), <frame_system::Pallet<T>>::block_number() $(, $values)*
+		)
+	};
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use crate::types::*;
+	use frame_election_provider_support::ElectionProvider;
+	use frame_support::pallet_prelude::*;
+	use frame_system::{pallet_prelude::*, RawOrigin};
+	use pallet_nomination_pools::PoolId;
+	use pallet_staking::Pallet as Staking;
+	use sp_runtime::{
+		traits::{Saturating, Zero},
+		DispatchResult,
+	};
+	use sp_staking::EraIndex;
+	use sp_std::{prelude::*, vec::Vec};
+	use weights::WeightInfo;
+
+	#[derive(scale_info::TypeInfo, codec::Encode, codec::Decode, codec::MaxEncodedLen)]
+	#[codec(mel_bound(T: Config))]
+	#[scale_info(skip_type_params(T))]
+	pub struct MaxChecked<T: Config>(sp_std::marker::PhantomData<T>);
+	impl<T: Config> frame_support::traits::Get<u32> for MaxChecked<T> {
+		fn get() -> u32 {
+			<T as pallet_staking::Config>::BondingDuration::get() + 1
+		}
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config:
+		frame_system::Config
+		+ pallet_staking::Config<
+			CurrencyBalance = <Self as pallet_nomination_pools::Config>::CurrencyBalance,
+		> + pallet_nomination_pools::Config
+	{
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>>
+			+ IsType<<Self as frame_system::Config>::RuntimeEvent>
+			+ TryInto<Event<Self>>;
+
+		/// The amount of balance slashed per each era that was wastefully checked.
+		///
+		/// A reasonable value could be `runtime_weight_to_fee(weight_per_era_check)`.
+		type SlashPerEra: Get<BalanceOf<Self>>;
+
+		/// The origin that can control this pallet.
+		type ControlOrigin: frame_support::traits::EnsureOrigin<Self::Origin>;
+
+		/// The weight information of this pallet.
+		type WeightInfo: WeightInfo;
+	}
+
+	/// The current "head of the queue" being unstaked.
```

Comment:
This basically means that `unstaking` is done sequentially one at a time, right? is there a reason for that?

Reply:
Ah, I can see that we want to limit ourselves to x eras per block, 1 user at a time to make sure we don't burn too much resources at the same time. 
It could, however, take ages to process the queue due to that, making `fast-unstake` not-so-fast and also not free.

Reply:
With something like 8 eras per block, we can unstake a nominator every 4-5 blocks, not so slow.